### PR TITLE
Portable model

### PR DESCRIPTION
**Pull request by philschatz**
I tried to add a few book validations from the errors we encountered last week with the vendor using the editor (Pages with duplicate UUID, chapters with the same name, modules in the book twice) but had trouble wrapping my head around all the caching & async communication going on.

This changes the language server in the following ways:

- disconnects the book model from VSCode, Promises, the filesystem reads/writes, and Caches.
- there is one place to update the model and the user gets immediate validation errors (or a list of nodes that need to be loaded first)
- There is one instance of every file in the model
- If a file does not exist (like for broken links) it is still instantiated but `.isEmpty()` is true

All the filesystem/vscode parts were moved into `model-adapter.ts`.

Also, there are some changes to the Job ~~Queue~~ Stack:
- The server enqueues the initial full-book load onto a "slow" ~~queue~~ _stack_
- Whenever a file is opened a job is added to the "fast" ~~queue~~ _stack_
- Whenever validation requires loading other files (like the link target page, or a target image) those get added to the stack and then the validation re-reports (another job on the stack)

With these changes, adding a validation message requires the following 3 pieces of information:

1. a message string that is shown to the user
2. a set of `Fileish` objects that need to be loaded before the validation can complete (like link target pages)
3. a set of `Range` objects that fail validation

```js
public getValidationChecks(): ValidationCheck[] {
    const imageLinks = this.imageLinks()
    return [
        {
            message: 'Missing image',
            nodesToLoad: imageLinks.map(l => l.image),
            fn: () => imageLinks.filter(img => !img.image.exists())
        }
    ]
}
```

Once all the dependent nodes are loaded, the Job Stack will re-run the validation check and report any errors that occur.

## Quick validation

There are 3 ways validation is faster:

Every time the model is updated we immediately get back either validation errors for that file or other files to load first. This way we do not have to load the whole book (except for the UUID-uniqueness check).

Also, since the model does not know about the filesystem we can immediately send it unsaved documents and get validation errors.

Third, the job runner has a fast/slow queue and open files are added to the fast queue.

Refs openstax/cnx#1577

# Commandline validator

As a demo, you can validate a book from the commandline. See `server/src/model/_cli.ts`:

```bash
npx ts-node@10.1.0 ./server/src/model/_cli.ts /path/to/book/repo

# This directory contains:
#   Books 4
#   Pages 138
#   Images 2346
#
# Validation Errors: 8
# ../osbooks-college-algebra-bundle/collections/college-algebra.collection.xml 13:6 Duplicate chapter title
# ../osbooks-college-algebra-bundle/collections/college-algebra.collection.xml 25:6 Duplicate chapter title
# ../osbooks-college-algebra-bundle/collections/college-algebra.collection.xml 10:4 Duplicate page
# ../osbooks-college-algebra-bundle/collections/college-algebra.collection.xml 11:4 Duplicate page
# ../osbooks-college-algebra-bundle/modules/m51240/index.cnxml 7:2 Duplicate Page/Module UUID
# ../osbooks-college-algebra-bundle/modules/m63490/index.cnxml 16:16 Missing image
# ../osbooks-college-algebra-bundle/modules/m63490/index.cnxml 14:8 Link target not found
# ../osbooks-college-algebra-bundle/modules/m63490/index.cnxml 7:2 Duplicate Page/Module UUID
```

# TODO

- [x] Remove old `.ts` files
- [x] :shirt: lint
- [x] Add model tests
- [x] Add adapter tests
- [x] Get to :100: % coverage
- [x] replace `.getFoo()` functions with `.foo` getters
- [x] add commandline validation demo
- [ ] Create a [Draft PR to add Quarx](https://github.com/openstax/poet/compare/portable-model...quarx) to push model updates to the client instead of request/response pairs

### CODE REVIEW

File: server/src/model-manager.spec.ts [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    expect(manager.allPages().size).toBe(0)
+    // Load the pages
+    const book = loadSuccess(first(loadSuccess(manager.bundle).books))
+    loadSuccess(first(book.pages))
+    expect(manager.allPages().size).toBe(1)
+  })
+  it('orphanedPages()', () => {
+    loadSuccess(first(loadSuccess(manager.bundle).books))
+    expect(manager.allPages().size).toBe(1)
+    expect(manager.orhpanedPages().size).toBe(0)
```
Comment (ckline-tryptic) [inline server/src/model-manager.spec.ts:61]: small spelling error

[the remark this answers is not in the document]
Comment (philschatz) [reply server/src/model-manager.spec.ts]: :blush: The downside of autocompletion. Fixed!

File: server/src/README.md [hunk 2]
```diff
@@ -1,21 +1,41 @@
...
+
+
+# Demo commandline validator
+
+To show/verify that the model works outside the language server, there is a CLI script that can validate a book repository.
+
+To run it:
+
+```bash
+npx ts-node@10.1.0 ./_cli.ts /path/to/book/repo /path/to/another/book/repo
```
Comment (therealmarv) [inline server/src/README.md:40]: minor: model directory missing

```npx ts-node@10.1.0 ./model/_cli.ts /path/to/book/repo /path/to/another/book/repo```

**Review by therealmarv — CHANGES_REQUESTED** [review]

File: server/src/README.md (hunk 2 above)
Comment (philschatz) [reply server/src/README.md]: Oops :blush: . Fixed!

**Review by therealmarv — APPROVED** [review]
For my point of view this are big changes (I mean that positive). I still have not wrapped my head around it fully but I like it what I see and I could test.

**Review by ckline-tryptic — APPROVED** [review]

File: server/src/model/_cli.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  const bits = fs.existsSync(n.absPath) ? fs.readFileSync(n.absPath, 'utf-8') : undefined
+  n.load(bits)
+}
+
+const pathHelper: PathHelper<string> = {
+  join: path.join,
+  dirname: path.dirname
+}
+
+;(async function () {
```
[the remark this answers is not in the document]
Comment (philschatz) [reply server/src/model/_cli.ts:30]: The self-invoked function is the only way to _sort of_ use `await` in the global scope. The semicolon is often optional but in this case JS would be confused and think that we broke the following expression onto multiple lines :shrug: `const pathHelper = {...}(...)`

image

File: server/src/README.md [hunk 4]
```diff
@@ -1,21 +1,41 @@
...
+1. no filesystem
+1. no async
+1. DOM is discarded after being parsed
+1. source location is remembered
+
+Each file is corresponds to a node in the model ([Bundle](./model/bundle.ts), [Book](./model/book.ts), [Page](./model/page.ts), Image).
+
+The Bundle is the root instance of a repository.
+
+Nodes are instantiated by a Factoryies on the Bundle.
```
[the remark this answers is not in the document]
Comment (philschatz) [reply server/src/README.md]: Thanks, fixed!

**Pull request merged**